deployer/helm_upgrade_decision.py

```diff
@@ -0,0 +1,367 @@
+"""
+Functions related to deciding which clusters and/or hubs need their *hub helm chart or
+support helm chart upgrading depending on an input list of filenames that have been
+added or modified in a GitHub Pull Request.
+"""
+import fnmatch
+import os
+from pathlib import Path
+from ruamel.yaml import YAML
+from rich.console import Console
+from rich.table import Table
+
+yaml = YAML(typ="safe", pure=True)
+
+# Determine if we are running a test or not. We set this env var to "test" in the
+# pyproject.toml file so it is set when the package is tested using pytest.
+test_env = os.getenv("RUN_ENV", False)
+
+
+def discover_modified_common_files(modified_paths: list):
+    """There are certain common files which, if modified, we should upgrade all hubs
+    and/or all clusters appropriately. These common files include the helm charts we
+    deploy, as well as the GitHub Actions and deployer package we use to deploy them.
+    Args:
+        modified_paths (list[str]): The list of files that have been added or modified
+            in a given GitHub Pull Request.
+    Returns:
+        upgrade_all_clusters (bool): Whether or not all clusters should be upgraded
+            since the support chart has changed
```

Comment:
_Original discussion:_ https://github.com/2i2c-org/infrastructure/pull/1093#discussion_r826210449

> I'm not confident the logic of upgrading all clusters and helm charts in them makes sense just because we change the support chart we deploy on all/most clusters, but that is probably matching the current behavior we have so 👍 for now!
> 
> There is a comment below:
> 
> ```
>     # If this filepath has changes, we should update the support chart on all clusters
>     support_chart_filepath = "helm-charts/support/*"
> ```
> 
> Hmmm, perhaps the `upgrade_all_clusters` was really `upgrade_support_chart_in_all_clusters` or similar?

Reply:
I suggest marking this as resolved without action.

It was related to decision logic, and I think as long as the decision logic made is easy to overview, whatever we do is fine.